drivers/mbox/mbox_stm32_hsem.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2024 Celina Sophie Kalus <hello@celinakalus.de>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/mbox.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+#include "stm32_hsem.h"
+
+#define LOG_LEVEL CONFIG_MBOX_LOG_LEVEL
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+LOG_MODULE_REGISTER(mbox_stm32_hsem_ipc);
+
+#define DT_DRV_COMPAT st_mbox_stm32_hsem
+
+#define HSEM_CPU1                   1
+#define HSEM_CPU2                   2
+
+#if DT_NODE_EXISTS(DT_NODELABEL(cpu0))
+#define HSEM_CPU_ID HSEM_CPU1
+#elif DT_NODE_EXISTS(DT_NODELABEL(cpu1))
+#define HSEM_CPU_ID HSEM_CPU2
+#else
+#error "Neither cpu0 nor cpu1 defined!"
+#endif
+
+#if HSEM_CPU_ID == HSEM_CPU1
+#define MBOX_TX_CHANNEL 0
+#define MBOX_TX_HSEM_ID CFG_HW_IPM_CPU2_SEMID
+
+#define MBOX_RX_CHANNEL 1
+#define MBOX_RX_HSEM_ID CFG_HW_IPM_CPU1_SEMID
+#else /* HSEM_CPU2 */
+#define MBOX_TX_CHANNEL 1
+#define MBOX_TX_HSEM_ID CFG_HW_IPM_CPU1_SEMID
+
+#define MBOX_RX_CHANNEL 0
+#define MBOX_RX_HSEM_ID CFG_HW_IPM_CPU2_SEMID
+#endif /* HSEM_CPU_ID */
+
+
+struct mbox_stm32_hsem_data {
+	const struct device *dev;
+	mbox_callback_t cb;
+	void *user_data;
+};
+
+static struct mbox_stm32_hsem_data stm32_hsem_mbox_data;
+
+static struct mbox_stm32_hsem_conf {
+	struct stm32_pclken pclken;
+} stm32_hsem_mbox_conf = {
+	.pclken = {
+		.bus = DT_INST_CLOCKS_CELL(0, bus),
+		.enr = DT_INST_CLOCKS_CELL(0, bits)
+	},
+};
+
+static inline void stm32_hsem_enable_rx_interrupt(void)
+{
+	const uint32_t mask_hsem_id = (1U << MBOX_RX_HSEM_ID);
+
+#if HSEM_CPU_ID == HSEM_CPU1
+	LL_HSEM_EnableIT_C1IER(HSEM, mask_hsem_id);
+#else /* HSEM_CPU2 */
+	LL_HSEM_EnableIT_C2IER(HSEM, mask_hsem_id);
+#endif /* HSEM_CPU_ID */
+}
+
+static inline void stm32_hsem_disable_rx_interrupt(void)
+{
+	const uint32_t mask_hsem_id = (1U << MBOX_RX_HSEM_ID);
+
+#if HSEM_CPU_ID == HSEM_CPU1
+	LL_HSEM_DisableIT_C1IER(HSEM, mask_hsem_id);
```

Comment:
I did it this way because in my opinion, it has a few benefits:

1. The hsem mask is always the same for each call, so encapsulating it makes slip-ups more unlikely
2. Code is more self-documenting because the function names are now more meaningful
3. When trying to look up the function, the extra step through the defines is avoided

It is how we have been implementing this at my company for a while now. If this coding style is not desirable here, however, I will of course change it.

Depending on what is decided in [the other thread](https://github.com/zephyrproject-rtos/zephyr/pull/69680#discussion_r1509252743), argument 1 might become obsolete, however.

Reply:
It is now done like in the IPM driver.